easycv/models/detection/dab_detr/dab_detr_transformer.py

```diff
@@ -0,0 +1,653 @@
+            if self.query_scale_type != 'fix_elewise':
+                if layer_id == 0:
+                    pos_transformation = 1
+                else:
+                    pos_transformation = self.query_scale(output)
+            else:
+                pos_transformation = self.query_scale.weight[layer_id]
+
+            # apply transformation
+            query_sine_embed = query_sine_embed[
+                ..., :self.d_model] * pos_transformation
+
+            # modulated HW attentions
+            if self.modulate_hw_attn:
+                refHW_cond = self.ref_anchor_head(
+                    output).sigmoid()  # nq, bs, 2
+                query_sine_embed[..., self.d_model //
+                                 2:] *= (refHW_cond[..., 0] /
+                                         obj_center[..., 2]).unsqueeze(-1)
+                query_sine_embed[..., :self.d_model //
+                                 2] *= (refHW_cond[..., 1] /
+                                        obj_center[..., 3]).unsqueeze(-1)
+
+            output = layer(
+                output,
+                memory,
+                tgt_mask=tgt_mask,
+                memory_mask=memory_mask,
+                tgt_key_padding_mask=tgt_key_padding_mask,
+                memory_key_padding_mask=memory_key_padding_mask,
+                pos=pos,
+                query_pos=query_pos,
+                query_sine_embed=query_sine_embed,
+                is_first=(layer_id == 0))
+
+            # iter update
+            if self.bbox_embed is not None:
+                if self.bbox_embed_diff_each_layer:
+                    tmp = self.bbox_embed[layer_id](output)
+                else:
+                    tmp = self.bbox_embed(output)
+                # import ipdb; ipdb.set_trace()
+                tmp[..., :self.query_dim] += inverse_sigmoid(reference_points)
+                new_reference_points = tmp[..., :self.query_dim].sigmoid()
+                if layer_id != self.num_layers - 1:
+                    ref_points.append(new_reference_points)
+                reference_points = new_reference_points.detach()
+
+            if self.return_intermediate:
+                intermediate.append(self.norm(output))
+
+        if self.norm is not None:
+            output = self.norm(output)
+            if self.return_intermediate:
+                intermediate.pop()
+                intermediate.append(output)
+
+        if self.return_intermediate:
+            if self.bbox_embed is not None:
+                return [
+                    torch.stack(intermediate).transpose(1, 2),
+                    torch.stack(ref_points).transpose(1, 2),
+                ]
+            else:
+                return [
+                    torch.stack(intermediate).transpose(1, 2),
+                    reference_points.unsqueeze(0).transpose(1, 2)
+                ]
+
+        return output.unsqueeze(0)
+
+
+class TransformerEncoderLayer(nn.Module):
```

Comment:
we should reuse these blocks from detr if we could

Reply:
done